test/hotspot/gtest/opto/test_rangeinference.cpp

```diff
@@ -0,0 +1,150 @@
+        ASSERT_EQ(U(hi), new_t.second._urange._hi);
+      } else {
+        ASSERT_EQ(std::numeric_limits<U>::min(), new_t.second._urange._lo);
+        ASSERT_EQ(std::numeric_limits<U>::max(), new_t.second._urange._hi);
+      }
+    }
+
+    {
+      U ulo = MIN2<U>(a, b);
+      U uhi = MAX2<U>(a, b);
+      TypeIntPrototype<T, U> t{{std::numeric_limits<T>::min(), std::numeric_limits<T>::max()},
+                               {ulo, uhi}, {0, 0}};
+      auto new_t = t.normalize_constraints();
+      ASSERT_TRUE(new_t.first);
+      DEBUG_ONLY(new_t.second.verify_constraints());
+      ASSERT_EQ(ulo, new_t.second._urange._lo);
+      ASSERT_EQ(uhi, new_t.second._urange._hi);
+      if (T(ulo) <= T(uhi)) {
+        ASSERT_EQ(T(ulo), new_t.second._srange._lo);
+        ASSERT_EQ(T(uhi), new_t.second._srange._hi);
+      } else {
+        ASSERT_EQ(std::numeric_limits<T>::min(), new_t.second._srange._lo);
+        ASSERT_EQ(std::numeric_limits<T>::max(), new_t.second._srange._hi);
+      }
+    }
+
+    {
+      U intersection = a & b;
+      U zeros = a ^ intersection;
+      U ones = b ^ intersection;
+      TypeIntPrototype<T, U> t{{std::numeric_limits<T>::min(), std::numeric_limits<T>::max()},
+                               {std::numeric_limits<U>::min(), std::numeric_limits<U>::max()}, {zeros, ones}};
+      auto new_t = t.normalize_constraints();
+      ASSERT_TRUE(new_t.first);
+      DEBUG_ONLY(new_t.second.verify_constraints());
+      ASSERT_EQ(zeros, new_t.second._bits._zeros);
+      ASSERT_EQ(ones, new_t.second._bits._ones);
+      ASSERT_EQ(ones, new_t.second._urange._lo);
+      ASSERT_EQ(~zeros, new_t.second._urange._hi);
+    }
+  }
+}
+
+template <class T, class U>
+static void test_normalize_constraints_random() {
+  constexpr int samples = 1000;
+  constexpr int parameters = 1000;
+  for (int i = 0; i < parameters; i++) {
+    T s1 = uniform_random<U>();
+    T s2 = uniform_random<U>();
+    T lo = MIN2(s1, s2);
+    T hi = MAX2(s1, s2);
+    U u1 = uniform_random<U>();
+    U u2 = uniform_random<U>();
+    U ulo = MIN2(u1, u2);
+    U uhi = MAX2(u1, u2);
+    U b1 = uniform_random<U>();
+    U b2 = uniform_random<U>();
+    U intersection = b1 & b2;
+    U zeros = b1 ^ intersection;
+    U ones = b2 ^ intersection;
+    TypeIntPrototype<T, U> t{{lo, hi}, {ulo, uhi}, {zeros, ones}};
+    auto new_t = t.normalize_constraints();
+    if (new_t.first) {
+      new_t.second.verify_constraints();
+    }
+    for (int j = 0; j < samples; j++) {
+      T v = uniform_random<U>();
+      if (!new_t.first) {
+        ASSERT_FALSE(t.contains(v));
+      } else {
+        ASSERT_EQ(t.contains(v), new_t.second.contains(v));
+      }
+    }
+  }
+}
+
+TEST_VM(opto, normalize_constraints) {
+  test_normalize_constraints_simple<jint, juint>();
+  test_normalize_constraints_simple<jlong, julong>();
+  test_normalize_constraints_random<jint, juint>();
+  test_normalize_constraints_random<jlong, julong>();
+}
```

Comment:
I would appreciate it if there were some explicit examples with explicit result verification. Just to make sure the methods are not systematically wrong in some silly way.

Reply:
My idea is that it is what `test_normalize_constraints_simple` would do, but I think adding some more explicit cases would help, too.

Reply:
Maybe you should just grep over the whole diff and look for `normal` to catch any remaining cases where it should be named `canonical` ;)